src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -243,12 +243,22 @@ private class DocumentSynchronizingContext
                     RequiredHostDocumentVersion = requiredHostDocumentVersion;
                     RejectOnNewerParallelRequest = rejectOnNewerParallelRequest;
                     _onSynchronizedSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
                     _cts = CancellationTokenSource.CreateLinkedTokenSource(requestCancellationToken);
 
-                    // This cancellation token is the one passed in from the call-site that needs to synchronize an LSP document with a virtual document.
-                    // Meaning, if the outer token is cancelled we need to fail to synchronize.
-                    _cts.Token.Register(() => SetSynchronized(false));
-                    _cts.CancelAfter(timeout);
+                    // If requestCancellationToken is already cancelled then registering
+                    // the callback will cause _cts to throw. So let's just set NotSynchronized now.
+                    if (_cts.IsCancellationRequested)
+                    {
+                        SetSynchronized(false);
+                    }
+                    else
+                    {
+                        // This cancellation token is the one passed in from the call-site that needs to synchronize an LSP document with a virtual document.
+                        // Meaning, if the outer token is cancelled we need to fail to synchronize.
+                        _cts.Token.Register(() => SetSynchronized(false));
```

Comment:
If registering this can potentially throw if the underlying token is cancelled we'll want to try/catch around it and SetSynchronized false in the fail case. Reason being is that the cancellation can happen at any time / be racey IIRC

Reply:
Ohh, I like that actually, I can replace the if with just the try-catch that w e have to have anyway